packages/codegen/src/commands.ts

```diff
@@ -0,0 +1,338 @@
+    const directory = await versionFileDirectoryPath(options);
+    const serviceTypesRecord: Record<string, string[]> = {};
+
+    const services = await this.listConfiguredServices(directory);
+    for (const service of services) {
+      serviceTypesRecord[service] = await this.generateForSingleService({
+        ...options,
+        service: service,
+      });
+      console.info(`generated types for ${service}`);
+    }
+
+    // identify shared resources and use the global versionfile to generate them
+    // this aviods that a local outdated versionfile may enforce outdated shared types
+    const sharedTypes = await this.generateSharedTypes(
+      serviceTypesRecord,
+      directory
+    );
+    const [sharedTypesExportStatement] = this.createExportStatementForPackage(
+      sharedTypes,
+      []
+    );
+    const blacklist = [...sharedTypes];
+    const exportStatements: string[] = [sharedTypesExportStatement];
+    console.info(`generated ${sharedTypes.length} shared types for package`);
+
+    for (const service of services) {
+      const types = serviceTypesRecord[service];
+      const [statement, skipped] = this.createExportStatementForService(
+        service,
+        types,
+        blacklist
+      );
+      exportStatements.push(statement);
+      blacklist.push(...types);
+      console.info(`${service}: skipped exporting [${skipped.join(', ')}]`);
+    }
+
+    const exportFilePath = path.resolve(directory, 'exports.gen.ts');
+    const fileContent = exportStatements.join('\n');
+
+    try {
+      await fs.writeFile(exportFilePath, fileContent);
+    } catch (error) {
+      throw new Error(
+        'unable to update import file for generated types: ' + error
+      );
+    }
+  };
+
+  private generateSharedTypes = async (
+    serviceTypesRecord: Record<string, string[]>,
+    packageDirectory: string
+  ): Promise<string[]> => {
+    const types = Object.values(serviceTypesRecord).reduce(
+      (acc, v) => acc.concat(v),
+      []
+    );
+
+    const sharedTypes: string[] = [];
+    for (let i = 0; i < types.length - 1; i++) {
+      const remain = types.slice(i + 1);
+      const t = types[i];
+      if (remain.includes(t)) {
+        sharedTypes.push(t);
+      }
+    }
+
+    const vfm = new VersionFileManager({ directory: packageDirectory });
+    const gen = new CodeGen(await vfm.read(), {
+      outputDir: packageDirectory,
+      autoNameInlinedRequest: false,
+
+      // rest is irrelevant - TODO: update types
+      service: '?',
+      version: '?',
+      scope: 'local',
+      filter: {
+        path: PassThroughFilter,
+      },
+    });
+
+    const generatedTypeNames = await gen.generateTypesFromSchemas(
+      (schemaName) => sharedTypes.includes(schemaName)
+    );
+    return generatedTypeNames;
+  };
+
+  private createExportStatement = (
+    fromDirectory: string,
+    types: string[],
+    blacklist: string[]
+  ): [string, string[]] => {
```

Comment:
I'd recommend returning an object rather than array. It helps a lot to understand what data there is without look at the implementation.

Something like:

```ts
private createExportStatement = (
    fromDirectory: string,
    types: string[],
    blacklist: string[]
  ): { code: string, skipped: string[] } => {
    // ...
  }

// and then
const { code } = createExportStatement(...)
```